File: 2023/day02/zig/solution.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const file_size = (try file.stat()).size;
    const buffer = try allocator.alloc(u8, file_size);
    defer allocator.free(buffer);

    _ = try file.readAll(buffer);

    // Trim trailing whitespace
    const input = std.mem.trim(u8, buffer, &std.ascii.whitespace);

    // Parse and solve
    const part1_result = try part1(allocator, input);
    const part2_result = try part2(allocator, input);

    std.debug.print("Part 1: {d}\n", .{part1_result});
    std.debug.print("Part 2: {d}\n", .{part2_result});
}

fn part1(_: std.mem.Allocator, input: []const u8) !u32 {
    var sum: u32 = 0;
    var lines = std.mem.splitSequence(u8, input, "\n");

    while (lines.next()) |line| {
        if (line.len == 0) continue;

        // Parse game ID
        const colon_pos = std.mem.indexOf(u8, line, ":") orelse continue;
        const game_id_str = line[5..colon_pos]; // Skip "Game "
        const game_id = try std.fmt.parseInt(u32, game_id_str, 10);

        // Parse reveals (semicolon-separated)
        const reveals_str = line[colon_pos + 2 ..]; // Skip ": "
        var reveals = std.mem.splitSequence(u8, reveals_str, ";");

        var possible = true;
        while (reveals.next()) |reveal| {
            // Parse each cube count in this reveal
            var cubes = std.mem.splitSequence(u8, std.mem.trim(u8, reveal, " "), ",");

            while (cubes.next()) |cube| {
                const trimmed = std.mem.trim(u8, cube, " ");
                var parts = std.mem.splitSequence(u8, trimmed, " ");

                const count_str = parts.next() orelse continue;
                const color = parts.next() orelse continue;

                const count = try std.fmt.parseInt(u32, count_str, 10);

                // Check against limits: 12 red, 13 green, 14 blue
                if (std.mem.eql(u8, color, "red") and count > 12) {
                    possible = false;
                } else if (std.mem.eql(u8, color, "green") and count > 13) {
                    possible = false;
                } else if (std.mem.eql(u8, color, "blue") and count > 14) {
                    possible = false;
                }
            }
        }

        if (possible) {
            sum += game_id;
        }
    }

    return sum;
}

fn part2(_: std.mem.Allocator, input: []const u8) !u32 {
    var sum: u32 = 0;
    var lines = std.mem.splitSequence(u8, input, "\n");

    while (lines.next()) |line| {
        if (line.len == 0) continue;

        // Parse game
        const colon_pos = std.mem.indexOf(u8, line, ":") orelse continue;
        const reveals_str = line[colon_pos + 2 ..]; // Skip ": "

        // Track minimum cubes needed for each color
        var min_red: u32 = 0;
        var min_green: u32 = 0;
        var min_blue: u32 = 0;

        // Parse reveals (semicolon-separated)
        var reveals = std.mem.splitSequence(u8, reveals_str, ";");

        while (reveals.next()) |reveal| {
            // Parse each cube count in this reveal
            var cubes = std.mem.splitSequence(u8, std.mem.trim(u8, reveal, " "), ",");

            while (cubes.next()) |cube| {
                const trimmed = std.mem.trim(u8, cube, " ");
                var parts = std.mem.splitSequence(u8, trimmed, " ");

                const count_str = parts.next() orelse continue;
                const color = parts.next() orelse continue;

                const count = try std.fmt.parseInt(u32, count_str, 10);

                // Update minimum needed for each color
                if (std.mem.eql(u8, color, "red")) {
                    min_red = @max(min_red, count);
                } else if (std.mem.eql(u8, color, "green")) {
                    min_green = @max(min_green, count);
                } else if (std.mem.eql(u8, color, "blue")) {
                    min_blue = @max(min_blue, count);
                }
            }
        }

        // Calculate power (product of minimums)
        const power = min_red * min_green * min_blue;
        sum += power;
    }

    return sum;
}

File: 2023/day02/go/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Draw struct {
	red   int
	green int
	blue  int
}

type Game struct {
	id    int
	draws []Draw
}

func parseGame(line string) Game {
	// Split "Game X: ..." into parts
	parts := strings.Split(line, ": ")

	// Extract game ID
	gameIDStr := strings.TrimPrefix(parts[0], "Game ")
	gameID, _ := strconv.Atoi(gameIDStr)

	// Parse draws (semicolon-separated)
	drawStrs := strings.Split(parts[1], "; ")
	draws := make([]Draw, 0, len(drawStrs))

	for _, drawStr := range drawStrs {
		draw := Draw{}
		// Parse each color count (comma-separated)
		cubes := strings.Split(drawStr, ", ")
		for _, cube := range cubes {
			parts := strings.Split(strings.TrimSpace(cube), " ")
			count, _ := strconv.Atoi(parts[0])
			color := parts[1]

			switch color {
			case "red":
				draw.red = count
			case "green":
				draw.green = count
			case "blue":
				draw.blue = count
			}
		}
		draws = append(draws, draw)
	}

	return Game{id: gameID, draws: draws}
}

func isGamePossible(game Game, maxRed, maxGreen, maxBlue int) bool {
	for _, draw := range game.draws {
		if draw.red > maxRed || draw.green > maxGreen || draw.blue > maxBlue {
			return false
		}
	}
	return true
}

func getMinimumCubes(game Game) (int, int, int) {
	minRed := 0
	minGreen := 0
	minBlue := 0

	for _, draw := range game.draws {
		if draw.red > minRed {
			minRed = draw.red
		}
		if draw.green > minGreen {
			minGreen = draw.green
		}
		if draw.blue > minBlue {
			minBlue = draw.blue
		}
	}

	return minRed, minGreen, minBlue
}

func part1(games []Game) int {
	sum := 0
	for _, game := range games {
		if isGamePossible(game, 12, 13, 14) {
			sum += game.id
		}
	}
	return sum
}

func part2(games []Game) int {
	sum := 0
	for _, game := range games {
		minRed, minGreen, minBlue := getMinimumCubes(game)
		power := minRed * minGreen * minBlue
		sum += power
	}
	return sum
}

func main() {
	// Read input file
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
		os.Exit(1)
	}

	// Parse games
	input := strings.TrimSpace(string(data))
	lines := strings.Split(input, "\n")

	games := make([]Game, 0, len(lines))
	for _, line := range lines {
		if line != "" {
			games = append(games, parseGame(line))
		}
	}

	// Solve both parts
	fmt.Printf("Part 1: %d\n", part1(games))
	fmt.Printf("Part 2: %d\n", part2(games))
}
